rfcs/0039-buffer-physical-layout.md

```diff
@@ -0,0 +1,667 @@
+with Allocate(x):
+    val = BufferLoad(x, [11, 37, 23, 101])
+
+# After applying the explicit reordering
+x = Buffer(name="x", shape=[16, 128/4, 64, 64, 4], reorder_splits=[], axis_separators=[])
+with Allocate(x):
+    val = BufferLoad(x, index=[11, floor(101/4), 37, 23, 101%4])
+
+# After flattening to 1-d
+x = Var(name="x")
+with Allocate(x, shape=[16 * (128/4) * 64 * 64 * 4]):
+    val = BufferLoad(x, index=[(128/4)*64*64*4*11 + 64*64*4*floor(101/4) + 64*4*37 + 4*23 + 101%4])
+```
+
+Lastly, an example of remapping from `NHWC` logical layout to `NCHWc`
+physical layout, packed into a 2-d physical layout with `NCH` in the
+first physical axis and `Wc` in the second physical axis.  This is the
+definition used by the current `"global.texture"` definition used for
+texture memory.  The change applied during SplitReorderIndices is
+identical to the previous example, but StorageFlatten produces a 2-d
+physical index.  The interpretation of this 2-d index depends on the
+target-specific codegen.
+
+```python
+# In TE schedule
+# s[x].transform_layout(lambda n,h,w,c: [n, c//4, h, te.AXIS_SEPARATOR, w, c%4])
+
+# Initial TIR graph
+attrs["buffer_layout_transformations"][x] = lambda n,h,w,c: [n, c//4, h, w, c%4]
+attrs["buffer_axis_separators"][x] = [2]
+x = Buffer(name="x", shape=[16,64,64,128])
+with Allocate(x):
+    val = BufferLoad(x, [11, 37, 23, 101])
+
+# After applying the explicit reordering.
+attrs["buffer_axis_separators"][x] = [2]
+x = Buffer(name="x", shape=[16, 128/4, 64, 64, 4])
+with Allocate(x):
+    val = BufferLoad(x, index=[11, floor(101/4), 37, 23, 101%4])
+
+# After applying StorageFlatten or FlattenBuffer.  The final result is
+# 2-d, due to the te.AXIS_SEPARATOR used in the `.transform_layout`.
+# The `"buffer_axis_separators"` attribute is set to [0], to
+# distinguish this 2-d flattened buffer from a 2-d unflattened buffer.
+attrs["buffer_axis_separators"][x] = [0]
+x = Var(name="x")
+with Allocate(x, shape=[16 * (128/4) * 64, 64 * 4]):
+    val = BufferLoad(x, index=[(128/4)*64*11 + 64*floor(101/4) + 37, 4*23 + 101%4])
+```
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This change may make it more difficult to reason about the memory
+layout when writing the `te.compute` definition.  When the physical
+layout differs from the logical layout, it isn't guaranteed that
+`A[i]` and `A[i+1]` will be adjacent.  For example, a tensor with
+compute definition defined in `NHWC` layout and with layout
+transformation to `NCHWc` defined by `[n, c//4, h, w, c%4]`, locations
+`(0,0,0,3)` and `(0,0,0,4)` in the compute definition will not be
+adjacent.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Can these design goals be met with existing features?
+
+  The `te.compute` function can be used to define an updated layout.
+  However, this introduces a new tensor that must be inlined to avoid
+  additional memory allocation, and cannot be used for input
+  parameters.
```

Comment:
Reading this, I realized that your proposal might solve the issue of N-d input `dense` / `batch_matmul` support discussed in https://github.com/apache/tvm/issues/8412 and https://discuss.tvm.apache.org/t/relay-nn-does-relay-nn-dense-supports-multi-dimensional-input/10343/7. The natural approach to enable N-d support is to wrap the existing 2-d compute / definition with `reshape`, and eliminate explicit `reshape` via `compute_inline`. 

So we will have `topi.reshape (N-d to 2-d) -> 2-d dense -> topi.reshape (2-d to N-d)`. While the first `reshape` can be trivially eliminated, the second one cannot (TensorIR has `reverse_compute_inline`, which may work for the second `reshape`). Currently,  `topi.reshape (N-d to 2-d) -> 2d dense -> topi.reshape (2-d to N-d)` pipeline does not work because TE bound inference fails to infer the precise bound for [the 2-d write buffer](https://github.com/apache/tvm/blob/d12e7259c2fc46c19fdff0aedfe543b506fb0441/python/tvm/topi/gpu/dense.py#L206) of `dense`. I don't fully understand the reason, but it seems the failure stems from the fact that the output buffer is N-d while the local write buffer is 2-d.  

Since the underlying physical memory layout doesn't change before and after reshape (it just collapses multiple "batch" dimensions into one), libraries like cudnn and mkl, that operates on pointers, supports N-d `dense` out of the box. So introducing explicit reshapes and eliminate them after-the-fact don't feel right. I've been bothered with this problem for some time, but `transform_layout` sounds like a perfect solution. Do you think `transform_layout` will enable supporting N-d `dense` and `batch_matmul` using existing 2-D schedules?

Note that the exact same problem arises in im2col based convolution: We need to "reshape" 2-D GEMM output of shape (N * H * W, C) to NHWC layout, but the physical layout doesn't change so explicit reshape should never be necessary. Overcoming this problem is a prerequisite for implementing tensorcore-based conv2d kernels based on [the implicit GEMM algorithm](https://github.com/NVIDIA/cutlass/blob/master/media/docs/implicit_gemm_convolution.md).

Reply:
I really like the application, and I think that it would be able to apply in this case with relatively small changes.  I had mainly been considering cases where the computation being performed is semantically identical, but with a different underlying data layout.  (e.g. conv2d_nhwc and conv2d_hwcn)  This is the reverse case, where the computation being performed is semantically different (e.g. "matrix multiply using indices 3/4 of a 4-d tensor" as opposed to "matrix multiply using indices 2/3 of a 3-d tensor"), but has an identical underlying data layout, and should share the same schedules.

Implementing it would have two steps.  The first would allow the `topi.nn.batch_matmul` to produce the different compute definitions for each tensor dimension, and the second would schedule any of those different compute definitions with the same shared function.

* Modifying the compute definition to accept N-d tensors.
  * Extracting the tensor shapes as `*other_dim, N, M = tensor.shape` instead of `batch, N, M = tensor.shape`.
  * Defining the compute definition as `lambda *indices: A[(*indices[:-2], indices[-2], k)] * B[(*indices[-2], k, indices[-1])]` instead of `lambda batch, i, j = A[batch, i, k] * B[batch, k, j]`.  (I made an improved parameter inspection for `.transform_layout` that, if extracted out to be used by `te.compute` could improve the readability to `lambda *other_indices, i, j = A[(*other_indices, i, k)] * B[(*other_indices, k, j)]`.)


* Modifying the schedule definition to accept N-d tensors.
  * Defining layout transformation function
    ```python
    def flatten_all_but_last_two(*other_indices, i, j):
        batch = 0
        for dim,index in zip(C.shape, index):
            batch = dim*batch + index
        return [batch, i, j]
    ```
  * Extracting the output tensor axes as `b, y, x = s[C].transform_layout(flatten_all_but_last_two)` instead of `b, y, x = s[C].op.axis`.

That said, as I was thinking through it, I don't think it strictly requires the `transform_layout` support in order to be implemented.  Since there's no change to the underlying data layout, generalizing the schedule could be done by changing the loop iteration order, which can already be done with `Stage.fuse`.
  
* Modifying the schedule definition to accept N-d tensors.
  * Extracting the output tensor axes as `*other_axes, y, x = s[C].op.axis` instead of `b, y, x = s[C].op.axis`.
  * Fusing the batch axes as `b = s[C].fuse(*other_axes)`.

Reply:
> That said, as I was thinking through it, I don't think it strictly requires the `transform_layout` support in order to be implemented.


Thanks, I've been thinking exclusively about using 2-d compute / schedules as they are with `reshape + inline`, so this simple solution didn't really occur to me. Partly because I was (1) lazy and (2) not sure if we can express "N-d indexing" with TE (like `A[(*indices[:-2], indices[-2], k)]` you showed above). I did try `b = s[C].fuse(*other_axes)` thing because that's required anyway for the reshape-based path as well, and that's where I hit the issue of imprecise bounds inference (second paragraph in my post above). 

I was making a wrong deduction that removing the need for explicit `reshape + inline` would let me workaround the limitation of TE bounds inference, and so `transform_layout` would "solve" this long-standing issue of mine. But I realized that these are distinct problems and that rewriting the compute and schedule for N-d wouldn't address the underlying problem. Thanks for helping me clearing my thoughts!